// @ohos/coverage/lib/src/commandLine/util/StackTrace.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StackTrace = void 0;
/**
 * StackTrace Model use to get stack trace instance
 */
class StackTrace {
    constructor(functionName, fileName, moduleName, row, column) {
        this._functionName = functionName;
        this._filePath = fileName;
        this._moduleName = moduleName;
        this._row = row;
        this._column = column;
    }
    get functionName() {
        return this._functionName;
    }
    set functionName(value) {
        this._functionName = value;
    }
    get filePath() {
        return this._filePath;
    }
    set filePath(value) {
        this._filePath = value;
    }
    get moduleName() {
        return this._moduleName;
    }
    set moduleName(value) {
        this._moduleName = value;
    }
    get row() {
        return this._row;
    }
    set row(value) {
        this._row = value;
    }
    get column() {
        return this._column;
    }
    set column(value) {
        this._column = value;
    }
}
exports.StackTrace = StackTrace;
